schemachanger: Implement `DROP CONSTRAINT` in declarative schema changer

Xiang-Gu opened the pull request:
This PR implements `ALTER TABLE t DROP CONSTRAINT cons_name` in declarative schema changer.

Supported constraints include Checks, FK, and UniqueWithoutIndex.

Dropping PK or Unique constraints will fall back to legacy schema changer, which in turn spits out an "not supported yet" error.

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96115)
<!-- Reviewable:end -->

Xiang-Gu committed:
schemachanger: Implement `DROP CONSTRAINT`

This commit implements `DROP CONSTRAINT`, turn it on by default, and
added tests.

Xiang-Gu committed:
schemachanger: Improve handling of FKs

This commit changed sc operations emmitted on the adding/dropping
path of FK constraint. The major change is to recognize that FK
constraints are "cross-descriptor" constraints and thus we need to
make accompanying changes in the referenced tables whenever we make
changes to the referencing table.

Xiang-Gu commented:
This is ready for a look!

postamar reviewed (commented):
I have only minor comments, besides: this needs some more test cases involving awkward scenarios such as dropping check constraints with type/sequence refs, partial unique-without-indexes, etc. They don't necessarily have to be end-to-end. Logic test cases are fine, provided they exist (do they?) also the data-driven tests in scplan might be useful to check that the planner does the right thing in those annoying multi-statement cases. I don't see the `cross-descriptor constraint is absent before *` rules show up anywhere in this PR.

ajwerner commented on pkg/sql/catalog/tabledesc/constraint.go:
```diff
@@ -430,6 +430,10 @@ func newConstraintCache(
 				c.all = append(c.all, fk)
 				c.fks = append(c.fks, fk)
+				if m.WriteAndDeleteOnly() {
+					c.allEnforced = append(c.allEnforced, fk)
```
One way to check on that would be to have something in the `end_to_end` test that uses and `EXPLAIN` on some select query that ensures we're not utilizing the constraint. 

For example:

```sql
CREATE TABLE t1 (i INT PRIMARY KEY);
CREATE TABLE t2 (i INT PRIMARY KEY REFERENCES t1(i));
```

```sql
EXPLAIN SELECT * FROM t2 join t1 on t2.i = t1.i LIMIT 1;
                                       info
-----------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: t1@t1_pkey
  │ equality: (i) = (i)
  │ equality cols are key
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 2 minutes ago)
        table: t2@t2_pkey
        spans: LIMITED SCAN
        limit: 1
(14 rows)
```

```sql
ALTER TABLE t2 DROP CONSTRAINT t2_i_fkey;
```
```sql
EXPLAIN SELECT * FROM t2 join t1 on t2.i = t1.i LIMIT 1;
                                         info
--------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • limit
  │ count: 1
  │
  └── • lookup join
      │ estimated row count: 1
      │ table: t1@t1_pkey
      │ equality: (i) = (i)
      │ equality cols are key
      │
      └── • scan
            estimated row count: 1 (100% of the table; stats collected 1 minute ago)
            table: t2@t2_pkey
            spans: FULL SCAN (SOFT LIMIT)
(16 rows)

```

Xiang-Gu replied:
done!

ajwerner commented on pkg/sql/schemachanger/scplan/internal/rules/current/helpers.go:
```diff
@@ -207,6 +207,14 @@ func isConstraint(e scpb.Element) bool {
 	return false
 }
 
+func isCrossDescriptorConstraint(e scpb.Element) bool {
```
use some words to explain this. Our taxonomy is arbitrary and could benefit from any sort of extra words we can sprinkle onto it.

Xiang-Gu replied:
done

postamar replied:
Also, these filters should have tests in `assertions_test.go`. Please don't skimp on these. They help keep this logic tight, reliable and future-proof.

Xiang-Gu replied:
I added a test there to assert that we have multiple partitions for constraints: is it cross-descriptor or not? Is it index-backed or not? etc.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_constraint.go:
```diff
@@ -0,0 +1,66 @@
+		RequiredPrivilege:   privilege.CREATE,
+	})
+	if constraintElems == nil {
+		// constraint not found but `IF EXISTS` is set
```
nit: make the comment a sentence.

While you're here, can we buffer a notice like postgres does?

```
postgres=# create table foo (i int primary key);
CREATE TABLE
postgres=# create table bar (i int primary key, j int references foo(i));
CREATE TABLE
postgres=# \d bar
                Table "public.bar"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 i      | integer |           | not null | 
 j      | integer |           |          | 
Indexes:
    "bar_pkey" PRIMARY KEY, btree (i)
Foreign-key constraints:
    "bar_j_fkey" FOREIGN KEY (j) REFERENCES foo(i)

postgres=# alter table bar drop constraint if exists bar_j_fkey;
ALTER TABLE
postgres=# alter table bar drop constraint if exists bar_j_fkey;
NOTICE:  constraint "bar_j_fkey" of relation "bar" does not exist, skipping
ALTER TABLE
```

ajwerner replied:
If it ends up being painful, can you file an issue to buffer such a notice?

Xiang-Gu replied:
no problem, this is rather easy. Added!

ajwerner commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_object.go:
```diff
@@ -48,7 +48,20 @@ func init() {
+			}
+		})
+
+	registerDepRule(
```
Can you comment that this is a weird rule that deals with the fact that these back-reference elements are bizarre? I think in practice we want to tie them to the name rather than to the descriptor to eliminate them.

Xiang-Gu replied:
done

ajwerner commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_object.go:
```diff
@@ -115,13 +128,27 @@ func init() {
 		},
 	)
 
+	registerDepRule(
```
Comment that we don't like these elements and this is to deal with that fact by tightly coupling them to the descriptor?

Can you take a stab at tightly coupling these to the name rather than to the descriptor and see if that just works?

Xiang-Gu replied:
done

ajwerner reviewed (commented):
Nothing major from me. This seems relatively painless. Great work!

Xiang-Gu commented on pkg/sql/catalog/tabledesc/constraint.go:
```diff
@@ -430,6 +430,10 @@ func newConstraintCache(
+				if m.WriteAndDeleteOnly() {
+					c.allEnforced = append(c.allEnforced, fk)
+					c.fksEnforced = append(c.fksEnforced, fk)
+				}
```
You are right and I tried that, but it breaks two existing unit tests ([test1](https://github.com/cockroachdb/cockroach/blob/0682851cda2cd29c7e9e033ec7e1c4d804ef5d06/pkg/sql/schema_changer_test.go#L5673) and [test2](https://github.com/cockroachdb/cockroach/blob/0682851cda2cd29c7e9e033ec7e1c4d804ef5d06/pkg/sql/schema_changer_test.go#L5769)) for legacy schema changer. Legacy schema changer has different semantics: the constraint does not move to being enforced until after we've backfilled the columns. It's an implicit state that we don't model (or, we model differently in the new schema changer). 
If we added the same thing for check constraints when building the cache, the inserts will fail because the check mutation is enforced while the unit test assumes that the insert should succeed (bc it's before the column is backfilled). 

I thought about deleting those two unit tests altogether but I think they are getting at something important so I didn't.

postamar replied:
Seems legit. Please add commentary to explain this.

Xiang-Gu replied:
already did!

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_constraint.go:
```diff
@@ -0,0 +1,66 @@
+) {
+	_, _, pie := scpb.FindPrimaryIndex(constraintElems)
+	if pie != nil {
+		panic(scerrors.NotImplementedError(t))
```
CRDB does not allow dropping PK constraint unless there is immediately followed by an ADD PRIMARY KEY.

I defer it to legacy schema changer for a few reasons: a). this is only possible when doing two schema changes in one transaction; I want to defer implementing this until we're more mature about reasoning/thinking/implementing/testing DDLs in transactions. b). The actual implementation could probably require more changes. E.g. How do we temporarily mark a PK as being "disabled" until a later ADD PRIMARY KEY in the same transaction alters the primary key; Consequently, this requires changing ADD PRIMARY KEY (nowadays, it only works if the current PK is the implicit rowId column).

Xiang-Gu replied:
I added more comments.

It's about reducing round-trips to put those punting logic for ADD CONSTRAINT in the map.

Xiang-Gu commented on pkg/sql/schemachanger/scexec/scmutationexec/constraint.go:
```diff
@@ -37,6 +37,19 @@ func (i *immediateVisitor) SetConstraintName(ctx context.Context, op scop.SetCon
+				inboundFK.ForeignKeyDesc().Name = op.Name
+				break
+			}
+		}
```
How is this possible? 

I assume what you wanted is that, in  `opgen_constraint_name.go`, you want to emit a different op based on the type of constraint. But element `ConstraintWithoutIndexName` only has the constraintID but we won't be able to know the constraint type inside `opgen_constraint_name.go`. 
Or, are you saying that we should introduce elements `CheckConstraintName`, `FKConstraintName`, etc. to achieve this?

postamar replied:
Oh, I'd forgotten about that. Never mind then! Sorry

Xiang-Gu commented on pkg/sql/schemachanger/scexec/scmutationexec/constraint.go:
```diff
@@ -434,23 +454,41 @@ func (i *immediateVisitor) MakeValidatedForeignKeyConstraintPublic(
+			if !foundInReferencedTable {
+				return errors.AssertionFailedf("failed to find accompanying inbound FK %v in"+
+					" referenced table %v (%v)", fk.Name, in.Name, in.ID)
+			}
```
Thanks for the advice! Done

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/current/helpers.go:
```diff
@@ -228,3 +236,11 @@ func isData(e scpb.Element) bool {
+		return true
+	}
+	return false
+}
```
My understanding is that when those two elements transition to absent, we drop the back-reference in the parent descriptor (hence the rule name). In that sense, those two elements are associated with (non)existence of the back-reference in the parent descriptor and thus I call them "backreferences". Perhaps a better name will be `IsParentDescriptorBackReference`?

postamar replied:
These elements model the forward reference that a descriptor holds to its parents. In the declarative schema changer, whatever owns the forward references also implicitly owns the back references. 
Here's an idea: just don't mention back-references at all in rules, call this filter `IsDescriptorParentReference` and rename the rules accordingly.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/current/helpers.go:
```diff
@@ -207,6 +207,14 @@ func isConstraint(e scpb.Element) bool {
+		return true
+	}
+	return false
+}
```
yes. 

In general, rule changes we made in this PR are to combat FKs: check constraints are way less "troubling" than FKs in that FKs can form cycles and that caused me a lot of headaches (think: t1->t2, t3->2, dropping t1,t2 cascade). I'm glad that checks don't behave like that, and that the existing rules works for checks even when it references other sequences/types. But I do want it to be principled so I'll add check constraint here.

I also added some logic test where we have cross-descriptor checks and dropping various descriptors in this case works as expected. Note that some of these operations are not even correctly supported in the legacy schema changer due to a known issue (check constraint references sequences by name, rather than by ID), but are correctly handled in the new schema changer.

Xiang-Gu commented on pkg/bench/rttanalysis/testdata/benchmark_expectations:
```diff
@@ -18,9 +18,9 @@ exp,benchmark
 15,AlterTableDropColumn/alter_table_drop_1_column
 15,AlterTableDropColumn/alter_table_drop_2_columns
 15,AlterTableDropColumn/alter_table_drop_3_columns
-9,AlterTableDropConstraint/alter_table_drop_1_check_constraint
-9,AlterTableDropConstraint/alter_table_drop_2_check_constraints
-9,AlterTableDropConstraint/alter_table_drop_3_check_constraints
```
I honestly don't. I'd love to understand it more afterward (maybe in the stability period?). I remember seeing similar results for other constraint-related statements in the new schema changer.

ajwerner reviewed (approved):
LGTM

Xiang-Gu commented:
CI failed due to a flaky test, borsing

bors r+
